language/stdlib/src/transaction_scripts.rs

```diff
@@ -1,74 +1,210 @@
 // Copyright (c) The Libra Core Contributors
 // SPDX-License-Identifier: Apache-2.0
 
-use crate::{compile_script, use_staged, MOVE_EXTENSION, STAGED_EXTENSION, TRANSACTION_SCRIPTS};
+//! Rust representation of a Move transaction script that can be executed on the Libra blockchain.
+//! Libra does not allow arbitrary transaction scripts; only scripts whose hashes are present in
+//! the on-chain script whitelist. The genesis whitelist is derived from this file, and the
+//! `Stdlib` script enum will be modified to reflect changes in the on-chain whitelist as time goes
+//! on.
+
+use crate::{
+    compile_script, use_staged, MOVE_EXTENSION, STAGED_EXTENSION, STAGED_OUTPUT_PATH,
+    TRANSACTION_SCRIPTS,
+};
+use anyhow::{anyhow, Error, Result};
 use include_dir::{include_dir, Dir};
+use libra_crypto::HashValue;
+use libra_types::transaction::SCRIPT_HASH_LENGTH;
 use once_cell::sync::Lazy;
-use std::path::PathBuf;
-
-// Encodes the absolute path for the transaction scripts directory.
-static TXN_SCRIPTS_DIR: Lazy<PathBuf> = Lazy::new(|| {
-    let mut path = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
-    path.push(PathBuf::from(TRANSACTION_SCRIPTS));
-    path
+use std::{convert::TryFrom, fmt, fs, path::PathBuf};
+
+/// Cache mapping each script to its compiled bytes. This allows clients to get the bytes/hash of a
+/// script without invoking the compiler or reading from disk.
+static ALL_SCRIPTS: Lazy<Vec<(StdlibScript, CompiledBytes)>> = Lazy::new(|| {
+    StdlibScript::all()
+        .iter()
+        .map(|script| {
+            let mut path = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
+            path.push(PathBuf::from(STAGED_OUTPUT_PATH));
+            path.push(PathBuf::from(TRANSACTION_SCRIPTS));
+            path.push(PathBuf::from(script.name()));
+            path.set_extension(STAGED_EXTENSION);
+            let contents = fs::read(path).unwrap();
+            (*script, CompiledBytes(contents))
+        })
+        .collect()
 });
 
-/// We include the compiled transaction scripts as binaries since the docker files do not copy any
-/// source files over. This could then cause issues when the client wishes to use a pre-packaged
-/// transaction script.
+// This includes the compiled transaction scripts as binaries. We must use this hack to work around
```

Comment:
@tzakian: do you have more context on this Docker problem? Blame attributes this comment to you and I want to make sure my changes don't bring the problem back...

Reply:
Oh boy.....I have stories to tell on this one (but at a later date). 

The long and short of it is that Docker only copies the binaries over. Because of this we need to make sure that the binaries for the stdlib/transaction scripts are built and stored in the Rust binary. So, if there's ever a place where you have "go to this directory and load the binary(/ies) from it" this will cause failures in the docker build.

Reply:
This is true but can be fixed by including these files in the docker build.sh. thought it it's one more additional overhead for maintaining scripts.

Reply:
Agreed. We had a discussion at that time and we decided to go the embedded-binary route for now. But I'd be fine to also see us copy the stdlib binaries over as well.